File: ABC087B/ABC087B.js
process.stdin.resume();
process.stdin.setEncoding("utf8");

var input_lines = []; //標準入力から受け取ったデータを格納する配列
var reader = require("readline").createInterface({
  //requireという処理でreadlineという機能を用いて標準入力からデータを受け取る
  input: process.stdin,
  output: process.stdout,
});

// ーーーー入力ーーーー

reader.on("line", (line) => {
  //line変数には標準入力から渡された一行のデータが格納されている
  input_lines.push(line); //ここで、input_lines配列に、標準入力から渡されたデータが入る
});

// ーーーー出力ーーーー
reader.on("close", () => {
  //受け取ったデータを用いて処理を行う
  const [yen500Num, yen100Num, yen50Num, yenSum] = input_lines;
  const FIVE_HUNDRED = 500;
  const ONE_HUNDRED = 100;
  const FIFTY = 50;
  let count = 0;
  for (let i = 0; i <= yen500Num; i++) {
    //答えゼロパターン：合計金額に持っている硬貨が足りない時
    if (yen500Num * FIVE_HUNDRED + yen100Num * ONE_HUNDRED + yen50Num * FIFTY < yenSum) {
      break;
    }
    //答えゼロパターン：合計金額が150円とかなのに50円持ってない時
    else if ((yenSum / 50) % 2 != 0 && yen50Num == 0) {
      break;
    } else {
      if (FIVE_HUNDRED * i > yenSum) {
        break;
      }
      let targetSum = yenSum - FIVE_HUNDRED * i;
      for (let j = 0; j <= yen100Num; j++) {
        if (ONE_HUNDRED * j > targetSum) {
          break;
        }
        if (targetSum - ONE_HUNDRED * j <= 50 * yen50Num) {
          count += 1;
        }
      }
    }
  }

  console.log(count);
});

File: ABC087B/ABC087B.py
def get_pattern_select_coins(yen500_num, yen100_num, yen50_num, yen_sum):

    FIVE_HUNDRED = 500
    ONE_HUNDRED = 100
    FIFTY = 50

    count = 0
    # 答えゼロパターン：合計金額に持っている硬貨が足りない時
    if (
        yen500_num * FIVE_HUNDRED + yen100_num * ONE_HUNDRED + yen50_num * FIFTY
        < yen_sum
    ):
        return 0
    # 答えゼロパターン：合計金額が150円とかなのに50円持ってない時
    elif yen_sum / 50 % 2 != 0 and yen50_num == 0:
        return 0
    else:
        for i in range(yen500_num + 1):
            if FIVE_HUNDRED * i > yen_sum:
                break
            remain500 = yen_sum - FIVE_HUNDRED * i
            for j in range(yen100_num + 1):
                if ONE_HUNDRED * j > remain500:
                    break
                # 500円と100円を引いた残額が50円で払えて、50の倍数の時カウントアップ
                remain100 = remain500 - ONE_HUNDRED * j
                if remain100 <= 50 * yen50_num:
                    if remain100 % 50 == 0:
                        count += 1
        return count
